PIG 16 - Gammapy package structure

cdeil opened the pull request:
We would like to discuss a few options to improve the Gammapy package structure for v0.14 (and v1.0 would be the same).

There are many options, but we expect that what will come out is that we do relatively small changes now, not a complete re-organisation. The discussion and document will still be valuable as a basis and reference for future developments.

Concretely, currently we propose to dissolve ``gammapy.background`` into other sub-packages.

We might add an overview / plan to clean up ``gammapy.utils`` here.
We don't want to discuss the big rename towards uniform class, function and parameter names in this PIG. That will come before v1.0, but a bit later, as another cleanup step after this PIG and several others that are open have been implemented.

@registerrier @adonath or anyone that wants to chime in at this early stage already (maybe @kosack or @cvaneldik with the HESS experience?) - please comment!

kosack commented:
>We propose to remove gammapy.background, and to move reflected.py to gammapy.spectrum, and the other few things to gammapy.image or gammapy.cube.

I would like to understand a bit the reasoning for moving background computations to the spectrum module.  Surely you can compute the background (even in 1D) for reasons other than computing a spectrum? It seems strange to mix those concepts.   

Perhaps that implies a bad splitting of images/cubes/lightcurves/spectra:  You could consider the approach of considering a spectrum as simply a cube with 1 spatial bin, 1 time bin, and multiple energy bins. In that case spectrum might be more appropriate in *cube*. Same with *image*: aren't images just cubes with 1 time and energy bin? It seems to be possible to merge all those into one structure. After all, they are all just "flux in a set of bins". That's perhaps a more drastic change, but it would clean up the code a lot I think, since computing a flux cube in time-energy-space is the same operation as a single lightcurve, or spectrum, or flux map, or combinations thereof (space-time, space-energy, time-energy)

adonath commented:
For maps we already gave up the distinction between the 2D (`SkyImage` in gammapy.image), 3D (`SkyCube` in gammapy.cube) and ND cases and moved to a uniform `gammapy.maps`. I think this was good change and we should continue in this direction. I agree with @kosack here, that if possible, we should try to give up the package separation in 1D, 2D, 3D analyse cases and rather structure it by classes / abstractions. So introduce `gammapy.models` with all the models (including base classes), introduce `gammapy.datasets` for all dataset classes (including base classes), and maybe even keep `gammapy.background` with all the 1D, 2D and 3D background estimation methods. I can see how this is a large proposed change, but if we don't do for 1.0, when is a better time then?

cdeil commented:
@kosack - Thanks for your comments!

I think for background-related code there's really the option to group it in `gammapy.background` or to have it in `gammapy.spectrum` for 1D background spectrum estimation methods that are usually regions-based like the `reflected.py` we have now, and for 3D maps based methods in `gammapy.cube`. Currently we have a mix, e.g. there's [gammapy.cube.make_map_background_irf](https://docs.gammapy.org/0.12/api/gammapy.cube.make_map_background_irf.html) or [gammapy.cube.models.BackgroundModel](https://docs.gammapy.org/0.12/api/gammapy.cube.models.BackgroundModel.html) or [gammapy.cube.MapMakerRing](https://docs.gammapy.org/0.12/api/gammapy.cube.MapMakerRing.html) in `gammapy.cube`, and e.g. [gammapy.background.RingBackgroundEstimator](https://docs.gammapy.org/0.12/api/gammapy.background.RingBackgroundEstimator.html) in `gammapy.background`. I don't see a clear distinction of what should go where, so naively I would think that dissolving `gammapy.background` and putting things more in one place for maps-based background estimation is a good idea. For reflected regions it's not easy to find a good home, it would have to go in `gammapy.spectrum` for now I guess, because `gammapy.cube` depends on `gammapy.spectrum`, and we should aim to avoid a circular dependency of spectrum and cube.

@kosack - Do you think keeping as-is or pulling all background-related code (namely the one I pointed out from `gammapy.cube`) into `gammapy.background` and to keep that is a good idea?

---

The suggestion to combine all data types of spectrum, images, cubes, lightcurves is probably a good one. I mention this in the first section in the `Outlook`. I just don't think we have any chance to get this done over the summer of 2019 and don't think we should change our plan and say "let's change everything and see how it goes, if we're not ready in the fall we'll do v1.0 sometime in 2020 or whenever ready". Doing this change for v2.0 aimed at fall 2020 sounds good to me. So this is mostly a comment for @adonath and @registerrier - think what realistically can be achieved, taking into account all the other work we still have to do.

cdeil commented:
Based on discussions at the Gammapy coding sprint, I've updated the proposal, and added a second Figure. We will discuss again today, and then I should be able to polish the text and finalise this proposal tomorrow.

cdeil commented:
@adonath @registerrier and anyone that has time - please do a final review of this PIG.
I would like to circulate it early next week.

Do I get a 🥇?

registerrier commented on docs/development/pigs/pig-016.rst:
```diff
@@ -0,0 +1,215 @@
+
+- move gammapy.utils.fitting to gammapy.modeling, and to move all models from
+  spectrum, image, cube and time to gammapy.modeling.models
+- dissolve gammapy.background, moving it's functionality into gammapy.spectrum
```
its functionalities

cdeil replied:
done

registerrier commented on docs/development/pigs/pig-016.rst:
```diff
@@ -0,0 +1,215 @@
+propose to expose them in gammapy.modeling.models (same as Astropy), and to
+introduce a factory function to make it easy to make models that is easy to
+import and use. E.g. it could be gammapy.modeling.make_model and take a model
+specification string or dictionary.
```
I would add a small code example of how this would work

cdeil replied:
I don't want to add a code example / prototype implementation of `make_model` here in the PIG. It's not something that can be done in a few lines, but I think it's clear that it can be done (that's what e.g. astropy-regions does from DS9 region strings, or what ctools does from model XML files, or what SkyCoord does from all kinds of strings that describe coordinates, or many other examples exists - this involves a parser and an object mapper.

registerrier replied:
Sorry for the confusion, I meant some example usages. I can see something like:
`pwl = make_model("powerlaw")`
But what if you want two? How do you do combined spectral and spatial?

cdeil replied:
@registerrier - Designing the model input specification (a "domain specific language") is not in scope here. What would we gain if I guess at something here? How does this affect this proposal? That will require a lot of discussion and falls into the model serialisation work.

I think what bothers you (rightly so) are the long model class names? Specifically you don't like that I propose we add "Spectral" to all spectral models to make this uniform?
https://docs.gammapy.org/0.12/spectrum/index.html#id4

Or do you not believe we will have a model factory function, and thus don't like that I claim here we can develop it?

cdeil replied:
I've added this guesswork for now:
```
Developing the syntax and schema for this
"model specification language" isn't in scope for this PIG. It will be done as
part of the effort on model serialisation in Gammapy in the coming months.
One could imagine that it uses YAML syntax like this::

    model = make_model("""
        type: spec, shape: powerlaw, pars: {norm:1e-12, index=2.7}
    """)

or a domain-specific language like this (a bit equivalent ot e.g. DS9 regions or
sky coordinates)::

    model = make_model("""
        spec.powerlaw(norm=1e-12, index=2.7)
    """)

or that one still uses Python syntax (not strings), only creates objects based
on attribute access, so calling it would look like this::

    model = Model.make.spec.powerlaw(norm="1e-12 cm-2 s-1 TeV", index=2.7)
```
Would prefer to remove it and only keep the sentence that it's not in scope to design this for this PIG, to avoid PIG review / discussion focusing on this, IMO mostly unrelated topic.
@registerrier @adonath  - Thoughts?

adonath replied:
I agree with @cdeil that proposing a solution for the model factory functions is not in scope for this PIG. I would not put any code at all. Just mention, that factory functions can be a solution to simplify imports of models for users.

registerrier replied:
OK. Right indeed this would drive the PIG too far off. At least it is good for me to see some examples or what you had in mind. OK not to show them, but keep what is before 'One could imagine...'

This is likely a long process, and won't be in for 1.0. Right? 
Then I would add that 'in the coming months, after v1.0 release'

cdeil replied:
That's what I put, and I added the explicit statement that it's not in scope and thus I'm not giving examples here, referencing model serialisation.

cdeil replied:
I don't want to guess as to what's in v1.0 or not here. I would hope that we have a first version of YAML serialisation in v0.14 and a better one in v0.15 and we would start using it also as input model spec language.

registerrier commented on docs/development/pigs/pig-016.rst:
```diff
@@ -0,0 +1,215 @@
+
+The motivation for this change is weak, but in discussions at the Erlangen
+coding sprint most Gammapy developers preferred it. The reasoning was that it's
+a bit easier to find and maintain models if they are all in one place.
```
I would mention here that the drawback is very long names, e.g. `SpectralPowerLaw` is probably still fine but `SpectralExponentialCutoffPowerLaw3FGL` is a real mess.

cdeil replied:
I've added the sentence "The drawback is that this generates very long class
names, such as e.g. SpectralExponentialCutoffPowerLaw3FGL."

registerrier commented on docs/development/pigs/pig-016.rst:
```diff
@@ -0,0 +1,215 @@
+The reflected region background code is not a great fit in gammapy/spectrum or
+anywhere in Gammapy except for gammapy.background, but moving it there does make
+sense if we say that all 1D spectrum code goes in gammapy.spectrum and all
+map-related analysis code goes in gammapy.cube.
```
Maybe add a sentence on 0D usage of cumulative statistics that should become part of 1D or 3D Dataset

cdeil replied:
Added this sentence:

> Scalar background statistics,
i.e. for a given region and energy band, can be created via methods on the 1D
and 3D dataset classes.

registerrier commented on docs/development/pigs/pig-016.rst:
```diff
@@ -0,0 +1,215 @@
+
+- asmooth.py and utils.py -> move to gammapy.cube
+- measure.py and profile.py -> replace with docs example or move to maps or cube
+- plotting.py -> replace with docs example or move to maps
```
Or add a plot sub package in the future?

adonath replied:
Make a decision here which functionality to keep and move and which to replace by docs examples? My proposal would be:

- Keep `ImageProfile` and `ImageProfileEstimator`, rename it to `MapProfile` and `MapProfileEstimator` and move it to `gammapy.maps` or `gammapy.cube`. This opens the way to extend it to arbitrary extra axes in future and implement e.g. likelihood fitting based algorithms.
- Remove `illustrate_colomap` and `grayify_colormap` from `image.plotting`, keep the other colormaps and the `MapPanelPlotter` but move it to `gammapy.maps.visualization`
- Move `ASmooth` to `gammapy.detect`, move `image.utils` to `maps.utils`
- Move `image.measure` to `maps.measure` for now, dissolve it maybe later.

cdeil replied:
@adonath - I've updated to this:
```
- utils.py -> one helper function, move to gammapy/maps/utils.py
- asmooth.py -> move to gammapy.detect
- measure.py and profile.py -> replace with docs example or move to maps or cube
- plotting.py -> move to gammapy/maps/plotting.py (remove illustrate_colormap and grayify_colormap)
```

For measure.py and profile.py we disagree at the moment. I think their usefulness and quality isn't good enough to keep, and we shouldn't spend time on it. Also it's not clear to me why you propose to put measure.py in gammapy.maps and profile.py in gammapy.cube, to me the code level seems very similar and naively I wouldn't expect one in one sub-package and the other somewhere else.

Would it be OK to keep this question in the PIG for now? It just means we don't have to agree on the solution now, and don't make these details part of the review by others, but it's something we will decide later on PR review.

cdeil replied:
@adonath - Changed to this following your suggestion:
```
- measure.py and profile.py -> move to gammapy.maps (and clean up)
```

registerrier commented on docs/development/pigs/pig-016.rst:
```diff
@@ -0,0 +1,215 @@
+
+gammapy.cube could be renamed to gammapy.map_analysis or gammapy.scripts to
+gammapy.analysis or similar changes depending on how their purpose and scope
+changes in the future datasets could be grouped in gammapy.datasets, if they are
```
purpose and scope change. In the future, datasets could be grouped in

cdeil replied:
fixed.

adonath commented on docs/development/pigs/pig-016.rst:
```diff
@@ -0,0 +1,215 @@
+introduced and we managed to support 2D and 3D maps with one container instead
+of separate gammapy.image.SkyImage and gammapy.cube.SkyCube.
+
+Now, if we move gammapy.image.models to gammap.modeling as proposed above, and
```
`gammap.modeling` -> `gammapy.modeling`

cdeil replied:
Fixed.

adonath commented on docs/development/pigs/pig-016.rst:
```diff
@@ -0,0 +1,215 @@
+Outlook
+=======
+
+During discussion of this PIG, several other ideas surfaced and were considered.
```
This paragraph is very chatty and I think statements like "either we will do it or not" are not really helpful inf focusing the development work. So  maybe shorten the paragraph to 2 sentences with a more concrete proposal? E.g.:

During discussion of this PIG, several other ideas surfaced. The following ideas should be considered and evaluated for feasibility (as done in `GH 2290`_)  in future:

cdeil replied:
done

adonath commented on docs/development/pigs/pig-016.rst:
```diff
@@ -0,0 +1,215 @@
+We propose to clean up gammapy.utils, e.g. gammapy/utils/coordinates.py still
+contains coordinate helper functions that were added before astropy.coordinates
+existed. We do not detail the cleanup here, this will be discussed on a
+case-by-case basis. After that we will review what is left and decide whether we
```
Maybe already list a preference here, based on the statement you've made above? Should it be a goal that users never import from `gammapy.utils`? I think this will help to focus the development process. Right now it reads like "either we will do it or not", which is completely open, but the statement about `very little` already suggests a direction I think.

cdeil replied:
Changed to

> We propose to clean up gammapy.utils, e.g. gammapy/utils/coordinates.py still
contains coordinate helper functions that were added before astropy.coordinates
existed. We do not detail the cleanup here, this will be discussed on a
case-by-case basis. After that we will review what is left and decide which
parts do mention in the public API docs and which to keep as purely internal code.

cdeil commented on docs/development/pigs/pig-016.rst:
```diff
@@ -0,0 +1,215 @@
+developments. We do expect that some of these changes will be made in the
+future (and some won't).
+
+Generalising maps to allow maps without a WCS and spatial axes, and / or using
```
I think having one container instead of different ones for spectra, cubes, time axes instead of several containers and datasets could change the package structure. And it was part of the discussion in this PIG (see earlier comment by Karl), and also with your idea to introduce `gammapy.datasets` - that need goes away if there's only 1-2 and they are Map-based, then they would just go in gammapy.maps. So I'll keep this sentence, but clarify: "It's not clear if and what changes this would imply for the Gammapy package structure."

cdeil commented:
@registerrier @adonath - Thanks for the feedback.

Most of it is addressed in c88fd2bc7 .
There are two unresolved conversations here concerning model factory and gammapy.image.
Please comment one more time; would be great to get this PIG out the stall today.

cdeil commented:
Final review period is over (see https://groups.google.com/d/msg/gammapy/DEbjzUFOe1o/_UsQsW4FDwAJ), didn't receive any further comments.

Changed PIG status and filled decision section, and did some minor RST formatting cleanup (zero content changes) in abe4a000b . Merging this now.

Thank you @adonath @registerrier @kosack and anyone who gave feedback on this PIG at the coding sprint in Erlangen!

I will file Github issues with the tasks to implement this and assign to the v0.14 milestone, to have the reminder and to discuss details if needed.

cdeil commented:
Reminder issues to implement this PIG: #2319 #2320 #2321 #2322
